### ISSUE TITLE
Support relative Layout values and view paths

### ISSUE THREAD

**Issue by mikeandersun** [aspnet/Mvc#3307]
I Have one controller action that return Views according to an input parameter like this code:

        [HttpGet]
        public async Task<IActionResult> RenderAViewInFoler(string foldername)
        {...
I can set the view path grammatically but my issue is setting the `layout.cshtml` and ` @Html.Partial("_footer")` that come from where ever `RazorViewEngine` returns.
I need to `layout.cshtml` and ` @Html.Partial("_footer")` come from the `foldername` path.
For temprory solution I set the path of those in View file like this:
`@{
     Layout = "~/foldername/_layout.cshtml";
}
`
FYI, in MVC 5 we could set full qualified name of layout for `View(...)` method.

**Comment by dougbu** [comment]
Could you give an example of a "fully qualified name" that doesn't look like `"~/foldername/_layout.cshtml"`?

**Comment by mikeandersun** [comment]
"~/folder1/_layout.cshtml"
"~/folder2/_layout.cshtml"
What all I needs is that when i set a path for a view like `View("~/folder1/View1.cshtml)"` then all partial view and the layout come from `folder1` not come from `Views/shared` or others place that searched by RazorViewEngine.
It seam RazorViewEngine  doesn't run per request to a controller action.

**Comment by dougbu** [comment]
#### tl;dr
The view and layout systems in MVC 5 and 6 are slightly different. You've found one of the differences -- `Controller` in MVC 6 lacks the `View(string viewName, string masterName)` overload available in MVC 5. In addition `Layout` values are not handled as relative paths as they are in MVC 5: They're either view names or fully-qualified paths (similar to `masterName` in the previous version).

We're going to look at adding relative path support throughout this system in MVC 6, providing the best of both words and avoiding odd searches that check for `/Views/Home/About.cshtml.cshtml`.

#### MVC 5 details
The MVC 5 lookup is approximately:

1. Find view page.
  1. If `viewName` passed to `View()` is `null`, search `~/Views/{controller}/{action}.cshtml` and `~/Views/Shared/{action}.cshtml`. (Area lookup checks `~/Areas/{area}/Views/{controller}/{action}.cshtml` and `~/Areas/{area}/Views/Shared/{action}.cshtml` first. I ignore areas from here on.) 
  2. Otherwise if `viewName` is fully-qualified, use given path.
  3. Otherwise, search `~/Views/{controller}/{viewName}.cshtml` and `~/Views/Shared/{viewName}.cshtml`. Note relative paths are not supported in this lookup.
2. Find all `_ViewStart.cshtml` files. Start in folder containing view page and move up directories until `~/Views/_ViewStart.cshtml` is checked.
3. Find `_Layout.cshtml` file.
  1. If `masterName` passed to `View()` is `null`, use `Layout` value (often set in `_ViewStart.cshtml`) instead.
    1. If `Layout` is `null`, stop here. Do not search for a default.
    2. If `Layout` is fully-qualified, use given path.
    3. Otherwise, if `Layout` is a relative path e.g. `_Layout.cshtml`, it is evaluated when the property is set relative to to the setter. E.g. if `~/Views/{controller}/_ViewStart.cshtml` contains `Layout = "_Layout.cshtml";`, use `~/Views/{controller}/_Layout.cshtml`.
  2. Otherwise if `masterName` is fully-qualified, use given path.
  3. Otherwise, search '~/Views/{controller}/{masterName}.cshtml' and '~/Views/{controller}/{masterName}.cshtml'. Note relative paths are not supported in this lookup.
4. Find partial views referenced from view page. This follows the same steps as 1. Again relative paths are not supported.

#### MVC 6 details
The MVC 5 lookup is approximately:

1. Find view page.
  1. If `viewName` passed to `View()` is `null`, search `~/Views/{controller}/{action}.cshtml` and `~/Views/Shared/{action}.cshtml`. (Area lookup checks `~/Areas/{area}/Views/{controller}/{action}.cshtml` and `~/Areas/{area}/Views/Shared/{action}.cshtml` first. I ignore areas from here on.) 
  2. Otherwise if `viewName` is fully-qualified, use given path.
  3. Otherwise, search `~/Views/{controller}/{viewName}.cshtml` and `~/Views/Shared/{viewName}.cshtml`. Note relative paths are not supported in this lookup.
2. Find all `_ViewStart.cshtml` files. Start in folder containing view page and move up directories until `~/Views/_ViewStart.cshtml` is checked.
3. Find `_Layout.cshtml` file.
  1. If `Layout` is `null`, stop here. Do not search for a default.
  2. If `Layout` is fully-qualified, use given path.
  3. Otherwise, search '~/Views/{controller}/{Layout}.cshtml' and '~/Views/{controller}/{Layout}.cshtml'. Note relative paths are not supported in this lookup.
4. Find partial views referenced from view page. This follows the same steps as 1. Again relative paths are not supported.

**Comment by dougbu** [comment]
Renaming to make our intended fix more clear. Fix will enable:
``` c#
View("~/Views/Folder/View.cshtml");    # fully-qualified paths work as they do today
View("View");                          # bare view names work at they do today (are searched for)
View("Folder/View.cshtml");            # relative paths (no leading ~ or / but end with .cshtml) added

Layout = ""~/Views/Folder/_Layout.cshtml"; # fully-qualified paths work as they do today
Layout = "_Layout";                        # bare view names work as they do today
Layout = "Folder/_Layout.cshtml";          # relative paths added
```

As with MVC 5's `Layout` property, relative paths are relative to the executing file when the property is set. `View("Folder/View.cshtml");` and `@Html.PartialAsync("Folder/View.cshtml")` are similar but there's no executing page in the first case so "Folder" is interpreted relative to the application root.

Note the last `View()` case is not supported in MVC 5 and is an enhancement for MVC 6. The current lack of `Layout = "Folder/_Layout.cshtml";` support in MVC 6 is a regression.

**Comment by mikeandersun** [comment]
When I have `View("Folder/View.cshtml");`
Is it possible to set `Layout = "_Layout.cshtml";` in `View.cshtml` and framework search `_Layout.cshtml` in the `Folder` directory?
And the same scenario for partials views?

**Comment by dougbu** [comment]
@mikeandersun once this bug is fixed, your example will find the view at `~/Folder/View.cshtml` and the layout at `~/Folder/_Layout.cshtml`. You'll get errors if either does not exist. Same for partials.

**Comment by mikeandersun** [comment]
When I have `View("Folder/View.cshtml");` :
What is important for me is haven't to set layout explicitly in the view and can set it like this :
`Layout = "_Layout.cshtml";` 
The `_Layout.cshtml` file is in the `Folder`
Sorry if i repeat one thing more and more :)

**Comment by dougbu** [comment]
Sorry, @mikeandersun what is your question? I have described how `Layout = "_Layout.cshtml";` will work.

**Comment by mikeandersun** [comment]
@dougbu it isn't a question ,only description of the scenario that I need.

**Comment by dougbu** [comment]
Commits 08dd77d through 95b3d2c pushed individually :frowning:

**Comment by dougbu** [comment]
@avanderhoorn note 3be6167 changes from the `isPartial` concept to `isMainPage`, updating property names in the `Microsoft.AspNet.Mvc.ViewFound` and `ViewNotFound` events to match.

Further the `Microsoft.AspNet.Mvc.Razor.BeginInstrumentationContext` and `EndInstrumentationContext` events no longer have `isPartial` properties. (The `isMainPage` concept is no longer relevant at the page level.)

**Comment by Tessalator** [comment]
I think being able to set the layout in the controller is important, especially in an enterprise setting and domain driven architectures. Here is an approach I use that is no longer supported.

I have an enterprise library of layouts (and other things that commonly live in shared). My project has business domain folders at the root. I created an attribute called "Layout". Now if I want to use a layout from the enterprise library I just decorate my controller with [Layout("~/Folder/_EnterpriseLayout.cshtml")] (I also have [LayoutPath("~/folder")], and [LayoutName("Name"). These are merged in a utility class. 

Another (to me) extremely valuable use of this is managing domain landing pages like "Index". I use the attributes at the action level too. When you enter a domain I want all of the home/index pages to have a common look, but the other pages have their own look. If I decorate the Index action with these attributes I can have just the home page have its own layout and all of the other actions use a different layout. In my example the [Layout...] on the controller.

I have derived a custom WebViewPage ("ViewPage") that "overrides" the View(0 methods - like,  new ActionResult View(view, layout, model){ base.View(view, layout, model) and handles all of the attribute fetching and resolution. This provides the ability to work with enterprise libraries.

**Comment by avanderhoorn** [comment]
@dougbu Thanks for the heads up!
